proposals/extension_get_contexts.md

```diff
@@ -0,0 +1,189 @@
+extensions will want to know what navigation state (`DocumentLifeCycle`) a
+`Context` is in, whether the frame’s `FrameType` is an outermost frame, and
+other fields like `parentDocumentId`, and `parentFrameId`. These are fields we
+would like to include in a future iteration of this API.
+
+### Asynchronous and Synchronous Version
+
+A potential avenue is that, since this new API is asynchronous whereas
+`getViews()` is synchronous, we could explore a synchronous version of this new
+API if there’s significant demand for it. That demand would need to justify why
+the usage of `await` on the existing asynchronous API isn't sufficient. As
+mentioned previously<sup>[1](#footnotes)</sup>, there are technical limitations
+that make it infeasible to synchronously fetch all `Context`s. We could provide
+two methods, one synchronous, and one asynchronous API methods:
+
+*   `getContexts()`: as defined above and asynchronous. It fetches all
+    `Context`s available in the process running for the profile.
+*   `getContextsSynchronous()` (name TBD): defined similarly to `getContexts()`,
+    but is synchronous. It returns the data for other `Context`s that are
+    available synchronously (running in the same process and on the same
+    thread). If called from a background script, it would likely only return the
+    background script `Context`.
+
+### Content Script Contexts
+
+[Content scripts](https://developer.chrome.com/docs/extensions/mv3/content_scripts/)
+run in a separate
+[Renderer](https://developer.chrome.com/blog/inside-browser-part3/) (and
+process) from the extension process. This proposal would not return those
+`Context`s since this adds significant complexity to the design in querying all
+Renderer processes in the browser to find content scripts. However, extensions
+could find it useful to have this `Context` metadata for places where the
+extension is executing JavaScript in a web page. It could be justified in the
+future if extension developers would have a significant need. If implemented the
+`frameUrl` field would be empty as providing a URL to the content script (within
+the extension) doesn't have a use case.
+
+### Messaging APIs Support `Context`s as Targets
+
+This new API now provides the ability to consider specifying targets via
+`Context`s to messaging APIs like `runtime.sendMessage()` or
+`tabs.sendMessage()`. With this support it would avoid
+[various workarounds](https://github.com/w3c/webextensions/pull/334#issuecomment-1343255899)
+needed to target specific pages with messages. `tabs.sendMessage()` accepting
+`Context`s would require work in retrieving content script `Context`s explained
+[above](#content-script-contexts).
+
+## Open Questions
+
+### Include `sandbox` Pages and Iframes as `Context`s?
+
+`sandbox` documents and iframes have an
+[opaque origin](https://html.spec.whatwg.org/multipage/browsers.html#concept-origin-opaque),
+should we return them? What if they come from within the extension (or outside
```

Comment:
IMO: no. Sandboxed documents have a `"null"` origin and don't have access to extension APIs. They're hardly different from regular sandboxed iframes from the web.